extension-manifest-v3/src/background/autoconsent.js

```diff
@@ -0,0 +1,92 @@
+/**
+ * Ghostery Browser Extension
+ * https://www.ghostery.com/
+ *
+ * Copyright 2017-present Ghostery GmbH. All rights reserved.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0
+ */
+
+import rules from '@duckduckgo/autoconsent/rules/rules.json';
+import { store } from 'hybrids';
+
+import Options from '/store/options.js';
+
+async function initialize(msg, tabId, frameId) {
+  const { dnrRules, autoconsent } = await store.resolve(store.get(Options));
+  const url = new URL(msg.url);
+
+  if (dnrRules.annoyances && !autoconsent.disallowed.includes(url.hostname)) {
+    const optOut =
+      autoconsent.all || autoconsent.allowed.includes(url.hostname);
+
+    chrome.tabs.sendMessage(
+      tabId,
+      {
+        action: 'autoconsent',
+        type: 'initResp',
+        rules,
+        config: {
+          enabled: true,
+          autoAction: optOut ? 'optOut' : '',
+          disabledCmps: [],
+          enablePrehide: optOut,
+          detectRetries: 20,
+        },
+      },
+      {
+        frameId,
+      },
+    );
+  }
+}
+
+async function evalCode(code, id, tabId, frameId) {
+  const result = await chrome.scripting.executeScript({
+    target: {
+      tabId,
+      frameIds: [frameId],
+    },
+    world: 'MAIN',
+    args: [code],
+    func: (code) => {
+      try {
+        return window.eval(code);
+      } catch (e) {
+        // ignore CSP errors
+        return;
+      }
+    },
+  });
+
+  chrome.tabs.sendMessage(
+    tabId,
+    {
+      action: 'autoconsent',
+      id,
+      type: 'evalResp',
+      result: result.result,
+    },
+    {
+      frameId,
+    },
+  );
+}
+
+chrome.runtime.onMessage.addListener((msg, sender) => {
+  if (msg.action !== 'autoconsent') return;
+
+  const tabId = sender.tab?.id;
```

Comment:
Maybe skip the following processing if "tabId" is undefined? tabId seems to be needed in "initialize" and "evalCode".